tv_apps/tv-deck/js/channel_manager.js

```diff
@@ -25,7 +30,305 @@
+    navigator.tv.getTuners().then(function onsuccess(tuners) {
+      this.isScanning = false;
+      this.currentTuners = {};
+      tuners.forEach(function initTuner(tuner) {
+        this.currentTuners[tuner.id] = {
+          tuner: tuner,
+          sources: {}
+        };
+      }.bind(this));
+
+      if (!this.playingTunerId) {
+        this.playingTunerId = tuners[0].id;
+      }
+
+      if (!this.playingTuner) {
+        this.fire('error');
+        return;
+      }
+
+      this.scanSources();
+    }.bind(this));
+  };
+
+  /**
+   * Retrieve all the currently available TV sources from current TV tuner.
+   */
+  proto.scanSources = function cm_scanSources() {
+    this.isScanning = true;
+    this.playingTuner.tuner.getSources().then(function getSources(sources) {
+      this.isScanning = false;
+      this.playingTuner.sources = {};
+      if (sources.length === 0) {
+        console.log('Error, no source found!');
+        return;
+      }
+
+      sources.forEach(function initSource(source) {
+        this.playingTuner.sources[source.type] = {
+          source: source,
+          channels: [],
+          channelIndexHash: {}
+        };
+      }.bind(this));
+
+      if (!this.playingSourceType) {
+        this.playingSourceType = sources[0].type;
+      }
+
+      if (!this.playingSource) {
+        this.fire('error');
+        return;
+      }
+
+      this.scanChannels();
+    }.bind(this));
+  };
+
+  /**
+   * Retrieve all the currently available TV channels from current TV source.
+   * We have to scan the channels before calling getChannels.
+   */
+  proto.scanChannels = function cm_scanChannels() {
+    this.isScanning = true;
+    var source = this.playingSource.source;
+    this.playingSource.channels = [];
+    this.playingSource.channelIndexHash = {};
+    if (source.isScanning) {
+      return;
+    }
+
+    var onscanningstatechanged = function(event) {
+      var state = event.state;
+      switch (state) {
+        case 'completed':
+        /* falls through */
+        case 'stopped':
+          this._onScanningCompleted();
+          source.removeEventListener(
+                                'scanningstatechanged', onscanningstatechanged);
+          break;
+        default:
+          break;
+      }
+    }.bind(this);
+
+    source.addEventListener('scanningstatechanged', onscanningstatechanged);
+
+    source.startScanning({
+      isRescanned: true
+    });
+  };
+
+  /**
+   * Retrieve all the currently available TV channels from current TV source.
+   * Sort channels for channel switching.
+   */
+  proto._onScanningCompleted = function cm_onScanningCompleted() {
+    this.playingSource.source.getChannels().then(function onsuccess(channels) {
+      this.isScanning = false;
+      if (channels.length === 0) {
+        console.log('Error, no channel found!');
+        return;
+      }
+
+      // Sort channels. Channel number can be XX-XX-XX
+      channels.sort(function(channelA, channelB) {
```

Comment:
We may use `channels.sort(this._compareChannel);` and change the arguments of _compareChannel to channel instead of number.